### ISSUE TITLE
Feature request: Macros saved in text files inside the project

### ISSUE THREAD

**Issue by joaofl** [texstudio-org/texstudio#54]
There is a feature I always missed in Texstudio, in the following context:
Someone is working in a given project, and program some macros, specially to replace commonly used symbols, which are hard to type (uC to turn into $\mu$C for example).

But the issue is that I have to program these macros on both my laptop and desktop. And everytime I format, have to reprogram it.

What it the macros were saved in the project working dir, so it could be synced (e.g. via dropbox), and yet the macros would work in the other machine.

Also because macros are very context specific, and sometimes some old macros end up bothering you on new projects.

Another COOL feature would be to able to put shortcuts to the programmed macros into the bookmarks on the left sidebar, with all the shortcuts. Or maybe simply a new tab showing the macros of that project.

I would gladly contribute if I can.

**Comment by benibela** [comment]
You can put scripts in the tex file:
```

% !TeX TXS-SCRIPT = oC
% //Trigger = oC
% editor.write("$\mu$C")
% TXS-SCRIPT-END 

```
(file needs to be closed and reopened to activate it)

**Comment by dbitouze** [comment]
@joaofl  As a side remark, you would be better to make use of the `siunitx` package for things like `$\mu$C`

```latex
\documentclass{article}
\usepackage{siunitx}

% We define a new unit: a micro coulomb
\DeclareSIUnit{\uc}{\micro\coulomb}
% We define a new command ("mec" for micro electric charge) for a physical
% quantity expressed in micro coulomb
\newcommand{\mec}[1]{\SI{#1}{\uc}} % 

\begin{document}
\begin{itemize}
\item \mec{5}
\item \si{\uc} stands for \num{e-6} coulomb.
\end{itemize}
\end{document}
```
![document](https://user-images.githubusercontent.com/1032633/37162899-bdf9eb9a-22f6-11e8-8e64-0e50cb75786a.png)

**Comment by joaofl** [comment]
@benibela I didnt know that. Im gonna try it now. It might do the job.

@dbitouze I've been declaring new commands every time, not with the siunitx package although. Most of the times it is just to make some keyboard shortcuts like, when I type: **fiG**, it turns into **Figure~\ref{fig:}**. And some other like those... More like a macro thing I would say...

**Comment by dbitouze** [comment]
>Most of the times it is just to make some keyboard shortcuts like, when I type: fiG, it turns into Figure~\ref{fig:}.

That's not incompatible: you could create a TeXstudio macro like the following one:

`Macros\0=Electric charge, \\ec{%<electric charge%>}, , eee`

that is:

- Name: Electric charge
- Trigger: eee
- LaTeX content: `\ec{%<electric charge%>}`

Then, just hit "eee" and that's it.

**Comment by joaofl** [comment]
@dbitouze 

> Macros\0=Electric charge, \\ec{%<electric charge%>}, , eee

where should this be declared?

```
% !TeX TXS-SCRIPT = oC
% //Trigger = oC
% editor.write("$\mu$C")
% TXS-SCRIPT-END 
```

I also tried to place the definition above on a tex file, but I did not execute. How shoudl this be done?
My idea is more to make this features more accessible to the laypeople.

**Comment by dbitouze** [comment]
@joaofl  It's what the TXS macro should looks like in `texstudio.ini` file after have been created as (not so precisely, sorry) specified: in TXS, menu Macros → Edit Macros → Add and then:

-  Name: Electric charge
-  Trigger: eee
-  LaTeX content: `\ec{%<electric charge%>}`

**Comment by joaofl** [comment]
@dbitouze  Ok, I see. Thanks for the clarification.

So my idea was to have some sort of per-project macro list, so we could sync it among different machines. How does that sound? I could try to implement that if it makes sense to more people other than myself.

**Comment by dbitouze** [comment]
Sounds interesting!

**Comment by benibela** [comment]
>I also tried to place the definition above on a tex file, but It did not execute. How should this be done?


You put it at the top of a tex file, close it, open it and write oC

**Comment by joaofl** [comment]
I started implementing this feature. I have doubt which is: when loading the macros from the local, and global config files, whats the best way of differentiating them? 

Should I add a checkbox to the macros manager for the user to defined whether its a local or global macro? Or do you have a better suggestion?

**Comment by timhoffm** [comment]
There are several things to do about macros. I don't think it's reasonable to add functionality like local macros to the current state of code. The whole system needs refactoring and adding functionality now will only make that more complicated.

- As a first step, it's reasonable to separate the global macros from texstudio.ini to a dedicated macros file.
- Then, there should be a more suited file format, probably either XML or JSON. To be discussed.
- It would be reasonable to store shortcuts with the macros. Currently, the shortcuts are bound to menu entries (and only implicitly to macros via their order in the macros menu). Ideally, one should be able to set shortcuts via the marco editor window.
- The whole macro editor window and code behind it could use a rewrite.
- There needs to be an upgrade path to migrate existing macros.

**Comment by joaofl** [comment]
@timhoffm 

>it's reasonable to separate the global macros from texstudio.ini to a dedicated macros file.
I also believe it makes sense. Although, would it still be part of the ConfiManager class?

In this sense, my first steps were to create the following 3 functions. One to convert from the old format to the new (XML, JSON...), one to load and populate the interface with the loaded macros, and another one to read the macros added/populated, and save them back to the file.

```
bool ConfigManager::convertMacros(const QString &in, const QString &out)
bool ConfigManager::readMacros(const QString &file)
bool ConfigManager::saveMacros(const QString &file)
```
As an argument, **read** and **save** functions get the file in which the macros are located. That is considering that it will be a separate file, which will no longer be part of the
`QSettings *config = persistentConfig;`

This will also make the way for having multiple macro files, including the local ones. How does it sound?

Another possibility the occurs to me is the possibility of having active **Python** macros... Many powerful applications have that feature, that adds great possibilities! It even open the doors for external contributions in the form of plugins. It also rises security concerns, which I believe can be addressed.

**Comment by sunderme** [comment]
I would discourage the implementation of python as script language. We provide txs for win,OSX and linux (quite a number of different distros) and I don't like to hassle by adding more dependencies which may be difficult to fulfill (e.g. windows)

**Comment by thatlittleboy** [comment]
Just adding to the conversation, I do like the idea of a revamped scripting interface for TXS (it's really one of the standout features, imo), but I am a little concerned with how conflicts in shortcuts between macros loaded from two different files would be resolved? Sorry if it's nothing constructive (I'm not knowledgeable in QT at all), but just wanted to raise potential issues and make them explicit in this discussion.

---------
Slightly off-topic @joaofl :

> Another COOL feature would be tohave the possibility of adding visual shortcuts to the macros into the bookmarks on the left sidebar, with all the other shortcuts. Or maybe simply a new tab showing the macros of that project. That would help to visually remember the macros you have on that project, and re-use them more often.

I just want to point out that it is already somewhat possible to do this (having "visual shortcuts to macros"), albeit with the "Central" Toolbar rather than the left "Structure" sidebar (where the symbols etc. are). Or realistically, with any other toolbars available to you.

**Adding macros to your toolbar:**
<img src="https://user-images.githubusercontent.com/30731072/37401847-62bcadc8-2781-11e8-9b85-4c3d021b4e3f.png" width=500>

**Changing default icon:**
<img src="https://user-images.githubusercontent.com/30731072/37401852-64a5d420-2781-11e8-933e-b20f605ce556.png" width=300>

**Result:**
<img src="https://user-images.githubusercontent.com/30731072/37401997-dabd552a-2781-11e8-8038-ddc72a9d3704.png" width=500>

**Comment by joaofl** [comment]
@thatlittleboy 
thanks for the highlight. I was not aware of that possibility actually. It is *almost how I idealized the macros menu, but not exactly. That because this requires some manual configuration to add a button to a single macro, while the menu on the left-most bar would open a panel with all the macros automatically populated. But it is nice to know!

To address the local X global macro loading issue, my idea was to add a flag the macro, that would specify whether its local or global. And at the time it is saved, its either saved on a file inside the project **my_macros.something**, or on the file inside **.config/texstudio/macros.something**.

At the time you open texstudio, it checks for both files to populate the list of macros. Should not be that complicated.

**Comment by thatlittleboy** [comment]
@joaofl 
Let's say I have `macro1` that I created on my **laptop** TXS, and then I set it with a shortcut of <kbd>Ctrl</kbd>+<kbd>Q</kbd> and flagged it as global. Then, while working on my **desktop** TXS, I created some other macro `macro2` which I flag as local (to the project), but also assigned to the *same* shortcut <kbd>Ctrl</kbd>+<kbd>Q</kbd>. (Suppose the desktop TXS has no `macro1`.)

Now what happens if I sync up the project (and thus the accompanying project macros) to my laptop? Should TXS keep the global `macro1` or local `macro2`? 

That's what I meant when I mentioned about conflict in the shortcuts. Or am I misunderstanding something here?

**Comment by joaofl** [comment]
@thatlittleboy 

I see your point. It makes very much sense. The conflicts are regarding the shortcuts...

I guess in this case we have to define a priority... Either hold the local or global one only. Probably the local has more importance, since it was created specifically to the project you are working on.

**Comment by thatlittleboy** [comment]
@joaofl Indeed. I'ld agree with that, I just wanted to raise the issue up explicitly. Thanks for working on this! :smile:

**Comment by joaofl** [comment]
@thatlittleboy but isnt it already possible to define different macros with the same shortcut? Not only the keyboard shortcut, but keywords as well. How does Texstudio handles it now? Never checked.

**Comment by thatlittleboy** [comment]
@joaofl Is that so? You mean keyboard shortcuts for macros? I just tried, TXS alerts me that the shortcut already exists for another command, and asks if I want to change.

Which ones point to the same shortcut? For "keywords", do you mean triggers?

**Comment by joaofl** [comment]
@thatlittleboy 

> For "keywords", do you mean triggers?
Exactly

**Comment by thatlittleboy** [comment]
@joaofl For triggers I think higher up macros have higher priority. (the ordering in the macro menu)

**Comment by joaofl** [comment]
@thatlittleboy yes... Just tested... In this sense, there should be means of moving them up and down, or just to make more explicitly what are the macros with higher priority.

**Comment by thatlittleboy** [comment]
@joaofl Excellent. I didn't realize this priority thing until you brought it up as well. 👍

**Comment by joaofl** [comment]
@thatlittleboy Actually the priorities are more like a natural consequence of how it is implemented. There is an array with all the macros/triggers loaded to it. As you type, texstudio checks that array from top to bottom, looking for any match. The first one to happen triggers the action and returns. Which seems to be ok.
